layers/cmd/schema.go

```diff
@@ -15,91 +15,204 @@
+		switch {
+		case len(ref.LayerID) > 0:
+			// Load the layer if not loaded before
+			_, loaded := layers[ref.LayerID]
+			if loaded {
+				break
+			}
+			b := ls.BundleByID{}
+			layer, err := b.LoadSchema(ref.LayerID)
 			if err != nil {
+				return err
+			}
+			layers[layer.GetID()] = layer
+
+		case ref.JSONSchema != nil:
+			_, loaded := entitiesMap[ref.JSONSchema.LayerID]
+			if loaded {
+				break
+			}
+			entity := jsonsch.Entity{
+				LayerID:   ref.JSONSchema.LayerID,
+				ValueType: variantType,
+				Ref:       ref.JSONSchema.Ref,
+			}
+			entitiesMap[entity.LayerID] = entity
+		}
+		return nil
+	}
+
+	// Load all layers, construct entities
+	for variantType, variant := range bundle.TypeNames {
+		if err := processRef(variantType, variant.BundleSchemaRef, schemaEntities); err != nil {
+			return nil, err
+		}
+		for _, ovl := range variant.Overlays {
+			if err := processRef(variantType, ovl, ovlEntities); err != nil {
 				return nil, err
 			}
-			_, err = b.Add(ctx, typeName, items[0], items[1:]...)
+		}
+	}
+
+	// If there are entities, import them
+	// First import schemas, then overlays
+	importJson := func(input map[string]jsonsch.Entity, typeTerm string) error {
+		importEntities := make([]jsonsch.Entity, 0, len(input))
+		for _, entity := range input {
+			importEntities = append(importEntities, entity)
+		}
+		if len(importEntities) > 0 {
+			jlayers, err := bundle.importJSONSchema(ctx, typeTerm, importEntities)
 			if err != nil {
-				return nil, err
+				return err
+			}
+			for k, v := range jlayers {
+				if _, exists := layers[k]; exists {
+					return fmt.Errorf("Multiple definitions for layer %s", k)
+				}
+				layers[k] = v
 			}
 		}
-		return &b, nil
+		return nil
+	}
+
+	if err := importJson(schemaEntities, ls.SchemaTerm); err != nil {
+		return nil, err
+	}
+	if err := importJson(ovlEntities, ls.OverlayTerm); err != nil {
+		return nil, err
 	}
 
-	b := ls.BundleByID{}
-	for id, layers := range bundle.Variants {
-		items, err := layers.Load(ctx, filepath.Dir(file))
+	resultBundle := ls.BundleByType{}
+	for variantType, variant := range bundle.TypeNames {
+		sch := layers[variant.GetLayerID()]
+		ovl := make([]*ls.Layer, 0, len(variant.Overlays))
+		for _, o := range variant.Overlays {
+			ovl = append(ovl, layers[o.GetLayerID()])
+		}
+		_, err := resultBundle.Add(ctx, variantType, sch, ovl...)
 		if err != nil {
 			return nil, err
 		}
-		_, err = b.Add(ctx, id, items[0], items[1:]...)
+	}
+	ret := make(map[string]*ls.Layer)
+	for variantId := range bundle.TypeNames {
+		layer, _ := resultBundle.LoadSchema(variantId)
+		if layer != nil {
+			ret[variantId] = layer
+		}
+	}
+	return ret, nil
+}
+
+func LoadBundle(ctx *ls.Context, file string) (ls.SchemaLoader, error) {
+	var bundle Bundle
+	if err := cmdutil.ReadJSON(file, &bundle); err != nil {
+		return nil, err
+	}
+	items, err := bundle.GetLayers(ctx, func(data string) (*ls.Layer, error) {
+		var input interface{}
```

Comment:
Loader func gets the filename, not the data. Read the file, then unmarshal json, and then unmarshal layer in the loader func.

Reply:
does this mean I read the file twice? Because I first read the file and unmarshal the contents into a Bundle struct. And then I call GetLayers on this bundle which should have all the Variants/TypeNames

Reply:
The func argument is the file name. So you should read the file whose name given in 'data', not the bundle.

Line 199 reads the bundle. The bundle refers to other files. This func loads those files.